File: OJ practices/LeetCode题目集/54. Spiral Matrix(medium).java
import java.util.ArrayList;
import java.util.List;

/*
 * @Author: SourDumplings
 * @Date: 2020-03-17 21:36:09
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/spiral-matrix/
 */

class Solution
{
    public List<Integer> spiralOrder(int[][] matrix)
    {
        List<Integer> res = new ArrayList<>();
        int m = matrix.length;
        if (m == 0)
        {
            return res;
        }
        int n = matrix[0].length;
        int[] limits = new int[]
        { n - 1, m - 1, 0, 0 };
        int size = m * n;
        int direction = 0;
        int i = 0, j = 0;
        for (int count = 0; count < size; ++count)
        {
            res.add(matrix[i][j]);
            switch (direction)
            {
            case 0:
            {
                // right
                if (j == limits[direction])
                {
                    ++limits[(direction + 3) % 4];
                    direction = (direction + 1) % 4;
                    ++i;
                }
                else
                {
                    ++j;
                }
                break;
            }
            case 1:
            {
                // down
                if (i == limits[direction])
                {
                    --limits[(direction + 3) % 4];
                    direction = (direction + 1) % 4;
                    --j;
                }
                else
                {
                    ++i;
                }
                break;
            }
            case 2:
            {
                // left
                if (j == limits[direction])
                {
                    --limits[(direction + 3) % 4];
                    direction = (direction + 1) % 4;
                    --i;
                }
                else
                {
                    --j;
                }
                break;
            }
            case 3:
            {
                // up
                if (i == limits[direction])
                {
                    ++limits[(direction + 3) % 4];
                    direction = (direction + 1) % 4;
                    ++j;
                }
                else
                {
                    --i;
                }
                break;
            }
            }
        }
        return res;
    }
}

File: OJ practices/LeetCode题目集/54. Spiral Matrix(medium).cpp
/**
 * @file 54. Spiral Matrix(medium).cpp
 * @author SourDumplings (changzheng300@foxmail.com)
 * @brief 
 * @version 1.0.0
 * @date 2024-04-06
 * 
 * @copyright Copyright (c) 2024 SourDumplings
 * 
 * https://leetcode.cn/problems/spiral-matrix/description/?envType=study-plan-v2&envId=top-interview-150
 * 
 */

class Solution
{
public:
    vector<int> spiralOrder(vector<vector<int>> &matrix)
    {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int> res;
        res.resize(m * n);
        int upperBound = 0, lowerBound = m - 1, leftBound = 0, rightBound = n - 1;
        char direction = 'r';
        int r = 0, c = 0;
        for (int i = 0; i < m * n; i++)
        {
            res[i] = matrix[r][c];
            switch (direction)
            {
            case 'r':
            {
                if (c == rightBound)
                {
                    direction = 'd';
                    ++upperBound;
                    ++r;
                }
                else
                {
                    ++c;
                }
                break;
            }
            case 'd':
            {
                if (r == lowerBound)
                {
                    direction = 'l';
                    --rightBound;
                    --c;
                }
                else
                {
                    ++r;
                }
                break;
            }
            case 'l':
            {
                if (c == leftBound)
                {
                    direction = 'u';
                    --lowerBound;
                    --r;
                }
                else
                {
                    --c;
                }
                break;
            }
            case 'u':
            {
                if (r == upperBound)
                {
                    direction = 'r';
                    ++leftBound;
                    ++c;
                }
                else
                {
                    --r;
                }
                break;
            }

            default:
                break;
            }
        }
        return res;
    }
};